highfidelity/hifi
# UI Pointer Optimization

- Generalized Pick and Pointer classes, managers, and scripting interfaces
- Stylus Pick and Pointer C++ classes
- Pointers (lasers and styluses) automatically generate pointer events
- OffscreenQmlSurface touch event improvements
- Updated scripts to use new system
- Fixes scrolling on tablet with lasers ([FB8076](https://highfidelity.fogbugz.com/f/cases/8076))
- Fixes stylus interaction with the tablet ([FB8896](https://highfidelity.fogbugz.com/f/cases/8896))
- Improves tablet interaction as game rate drops ([FB7920](https://highfidelity.fogbugz.com/f/cases/7920/), [FB8392](https://highfidelity.fogbugz.com/f/cases/8392/))

Test plan:
- All grabbing tests: https://highfidelity.testrail.net/index.php?/suites/view/20
- All tablet interaction tests: https://highfidelity.testrail.net/index.php?/suites/view/27
- All context overlays tests: https://highfidelity.testrail.net/index.php?/suites/view/103
- Haptics should behave the same as in master.  Namely: strong pulse on click with lasers/stylus, small pulse when hovering over virtual keyboard letters.
- Create some web entities and use the hand lasers on them.  You should be able to scroll/click, etc.  
- If you trigger or press with the stylus on the tablet, you should be able to scroll off the surface.
- Interactions with lasers and stylus on the tablet should be faster and less glitchy.  Try in an empty domain and a heavily loaded one where your game rate is much lower (dev-welcome is good, or anywhere where your game rate is less than ~45 in HMD).
- Try at different avatar scales.  Stylus and lasers should still work on the tablet.
- Run all the LaserPointer tests in the hifi_tests repository: https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/interaction/pointer/laser
